### ISSUE TITLE
Undo breaks after completion

### ISSUE THREAD

**Issue by tweekmonster** [Shougo/deoplete.nvim#346]
It seems `<Plug>(deoplete_start_complete)` breaks undo.

## Environment Information

* OS: Ubuntu 16.04.1 LTS
* Nvim: v0.1.6-167-g22dfe69

## Reproduction

- Use standard minimal init.
- Trigger completion (.e.g `ca` in a Vim script)
- When the completion menu is displayed, press `<esc>`
- Press `A` in normal mode to re-enter insert mode and trigger completion

For every time `<esc>A` is pressed, a new undo is created.

## Solution?

If I add `undojoin` above [line 17](https://github.com/Shougo/deoplete.nvim/blob/d53c8f46f1aff5e015076c86dbcdfcc35d5a9015/autoload/deoplete/mapping.vim#L17) in `mapping.vim`, the issue is resolved.  But, I'm not sure if there are cases where `undojoin` will cause an error.

```diff
diff --git a/autoload/deoplete/mapping.vim b/autoload/deoplete/mapping.vim
index a14dd1d..8c4cc1d 100644
--- a/autoload/deoplete/mapping.vim
+++ b/autoload/deoplete/mapping.vim
@@ -14,6 +14,7 @@ endfunction"}}}
 function! deoplete#mapping#_do_complete(context) abort "{{{
   if b:changedtick == get(a:context, 'changedtick', -1)
         \ && mode() ==# 'i'
+    undojoin
     call complete(a:context.complete_position + 1, a:context.candidates)
   endif
```

**Comment by tweekmonster** [comment]
[Here's a video of the problem.](https://share.esdf.io/lvjoYa9p0u/Screen-Recording-2016-09-29-10-28-37.mp4)

**Comment by tweekmonster** [comment]
@Shougo Sorry I just noticed a problem.  I think this needs a revert.  When I complete with a `.` after undo, the undo state cause `undojoin` to generate an error.  I'm not sure what the difference is, but maybe it's because `.` doesn't wait for a delay?

Maybe the context needs to include a variable to tell the `_do_complete()` function that it was a delayed trigger from the rplugin?

**Comment by Shougo** [comment]
> Maybe the context needs to include a variable to tell the _do_complete() function that it was a delayed trigger from the rplugin?

Please test the change.
I have not time to test it.

**Comment by tweekmonster** [comment]
So far, it seems that when `a:context.event` is `InsertEnter` it breaks undo.  But, `TextChangedI` raises an error if `undojoin` is used after performing an undo.  Do you have an idea of why the undo is broken durning an `InsertEnter` event?

**Comment by Shougo** [comment]
@tweekmonster
I havre reproduced your problem. I think it is feature.
Because, you can reproduce it by `<C-x><C-n>` instead of deoplete.
Vim's completion is implemented as text changes.
So, it breaks undo.

I have added `a:context.event` check.

**Comment by tweekmonster** [comment]
Thanks!

> I think it is feature.

Hmm, if it causes no change to the buffer it shouldn't add an empty undo item right?  Maybe it's not a bug, but I think it's an undesirable effect.

I just noticed that `ia<bs><esc>` (insert one character and delete before escape) causes an undo to be created even though the buffer is unchanged.  Also `cc<esc>` on an empty line.  I guess I never noticed since I don't do that often. I always assumed that `insert` -> `normal` only created undos if the buffer contents changed.

**Comment by Shougo** [comment]
It is created in `changed_common()`.
https://github.com/neovim/neovim/blob/master/src/nvim/misc1.c#L1973

**Comment by tweekmonster** [comment]
☹️  I'll have to study it later.

**Comment by tweekmonster** [comment]
@Shougo There still seems to be a problem.  In the examples below `vi` will complete to `vim`:

```
vi|
```

When you press `<esc>` a new undo will be created, and pressing `A` the completion reappears.  Then pressing `<c-y>` accepts `vim`:

```
vim|
```

**While still in insert mode** type: `<cr>vi`:

```
vim
vi|
```

Again from here press `<esc>A<c-y>` and it should be:

```
vim
vim|
```

But, when you press `u` both lines are deleted (empty buffer).  But, I expect it to be:

```
vi|
```

I checked `changed_common()` in `misc.c`, but I don't think that's the actual cause of this issue.  `changed_common()` calls `changed()` which *always* increments `b:changedtick`.  When I was testing this problem with `A<esc>A<esc>...`, `changenr()` increments, but `b:changedtick` doesn't.  I'm still investigating, but I'm if I'm looking in the right places.

`call complete()` is what increments `changenr()`.  The `undojoin` must come before it, and it must only happen after `InsertEnter`.  But, if that happens more than once without an actual change in the buffer, the extra `undojoin` will be wrong.  I'm not sure if this is an issue that can be solved in this plugin  ☹️

**Comment by Shougo** [comment]
It seems Vim's feature.
I recommend to use `deoplete#undo_completion()` to cancel the completion.

**Comment by tweekmonster** [comment]
> It seems Vim's feature.

Unacceptable feature!  I found the issue, but I'm not sure if it's the best solution: https://github.com/tweekmonster/neovim/commit/43d239d68f47260a1b4b134e3a3f9423a2eb1cdf

> I recommend to use `deoplete#undo_completion()` to cancel the completion.

That still won't fix the the problem.  `complete()` creates a new undo entry.  Since I'm using `noinsert` the undo entry is always blank.  Even if I use `inoremap <expr> <esc> pumvisible() ? '<esc>u' : '<esc>'` it would mean that I'll end up with a bunch of useless undo branches.  `deoplete#undo_completion()` would still create an undo entry, so it's not a good option to me.

**Comment by tweekmonster** [comment]
@Shougo Looks like [mapping.vim#L17-L19][1] should be reverted once `8.0.0041` is merged into Neovim 🎉 

[1]: https://github.com/Shougo/deoplete.nvim/blob/667504f4372968494c47fb880d4f18cf602edaa6/autoload/deoplete/mapping.vim#L17-L19

**Comment by tweekmonster** [comment]
After this PR is merged, the `undojoin` should be reverted: https://github.com/neovim/neovim/pull/5499

**Comment by Shougo** [comment]
OK
